### ISSUE TITLE
Don't enforce contiguous memory

### ISSUE THREAD

**Issue by kalzoo** [rigetti/qcs-sdk-rust#26]
Currently, in `qcs/src/qpu/mod.rs`, `organize_ro_sources`, there's a block which enforces that readout targets contiguous memory regions starting from `0`, that is, this program is invalid because it skips `ro[0..6]`:

```
DECLARE ro BIT[8]
MEASURE 7 ro[7]
```

However, this is both valid Quil and accepted by `pyQuil` and the QCS API. The `ro_sources` returned look like this:

```
ro_sources={<MRef q7_unclassified[0]>: 'q7_unclassified', <MRef ro[7]>: 'q7'}
```

and the execution result looks like this (note the uninitialized memory in indices 0..6:

```
readout_data={'ro': array([[-3458764513820540928, -3458755728090571410,  8314613781408776197, 7598826450156352046,    72057793442439982,   146087754952868096, 654311424, 0]])}
```

The assertion that this Rust SDK is making makes sense in terms of being able to iterate safely through the result, esp via C, but it cannot add constraints to the Quil language in order to do so. We could consider a map (probably a `BTreeMap` to preserve ordering for iteration) or a sparse Vec (maybe a `Vec<Option<_>>`) to store these sparse results.

**Comment by dbanty** [comment]
The trickier thing is going to be figuring out the C API for this. Do we need a function call for accessing each slot of data (e.g. `get_data(result.handle, "ro", 7)`)? I feel like that would add significant overhead in the run -> get results -> tweak params -> run loops.

We could allocate the whole array and just fill in 0s for any values we didn't get? What does `pyQuil` do?

**Comment by kalzoo** [comment]
> We could allocate the whole array and just fill in 0s for any values we didn't get? What does pyQuil do?

Yes, I believe that's what we should do, to present the facade of initialized-but-not-written memory being returned from execution. It appears that [pyQuil allocates but does not initialize a new array](https://github.com/rigetti/pyquil/blob/497791e8108d8780109d75410be786c5f6e590ea/pyquil/api/_qpu.py#L80), which I'd describe as an oversight and likely a bug.

**Comment by jselig-rigetti** [comment]
This is working as suggested now (pyquil depends on qcs-sdk-rust)

```py
from pyquil import get_qc, Program
from pyquil.gates import H, CNOT, MEASURE
from pyquil.quilbase import Declare

program = Program(
    Declare("ro", "BIT", 10),
    H(0),
    CNOT(0, 1),
    MEASURE(1, ("ro", 5)),
).wrap_in_numshots_loop(10)

qc = get_qc("Cepheus-1-108Q")
executable = qc.compile(program)
result = qc.run(executable)

ro = result.get_register_map()["ro"]
print(ro)
```

returns 
```
[[0 0 0 0 0 0]
 [0 0 0 0 0 1]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]]
```
